pkg/extensions/tbtc/recovery/derive_address.go

```diff
@@ -129,17 +129,27 @@ func deriveAddress(extendedPublicKey string, addressIndex uint32) (string, error
 // the specified index.
 func ResolveAddress(
 	beneficiaryAddress string,
+	storage *DerivationIndexStorage,
 	chainParams *chaincfg.Params,
 ) (string, error) {
-	//FIXME: pull the address index from persistence
-	addressIndex := uint32(0)
 	// If the address decodes without error, then we have a valid bitcoin
 	// address. Otherwise, we assume that it's an extended key and we attempt to
 	// derive the address.
 	_, err := btcutil.DecodeAddress(beneficiaryAddress, chainParams)
 	if err != nil {
-		return deriveAddress(beneficiaryAddress, addressIndex)
-	} else {
-		return beneficiaryAddress, nil
+		addressIndex, err := storage.GetNextIndex(beneficiaryAddress)
+		if err != nil {
+			return "", err
+		}
+		derivedAddress, err := deriveAddress(beneficiaryAddress, uint32(addressIndex))
+		if err != nil {
+			return "", err
+		}
+		err = storage.Save(beneficiaryAddress, addressIndex, derivedAddress)
+		if err != nil {
+			return "", err
+		}
```

Comment:
Should we `Save` the new address once we know the transaction was signed and broadcasted (or at least written to logs)? Right now, in case of errors from signing, we'll keep writing new addresses to the storage and those addresses will never be used.

Reply:
Ah, it is here because we want to make sure two keeps cannot use the same index for BTC recovery? I wouldn't care about it for now, to be honest. Liquidations are not happening that often and even with the current version there is still a risk in case two liquidations happen at the same time. Especially that in the current version we derive address (and store new index to disk, no matter if it's been used or not) in `DoWithDefaultRetry` loop that is potentially retried after each problem, including announce protocol not completed successfully.

Reply:
I moved the `Save` piece to after the broadcast call, so now we have a lower chance of persisting unused indexes